File: Polynomial Regression/polynomial_regression_saurabh.R
#Importing data
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Spliting the dataset into Training set & Test set
# Spliting the dataset into Training set & Test set
#We dont have enough data to split it into train and test set
#install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 0.8)
# training_set = subset(dataset, split == T)
# test_set = subset(dataset, split == F)

#Fitting Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

#Fitting Polynomial Regression to the dataset
dataset$Level2 = dataset$Level^2          #added new column which contains value level to the power of polynomial
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary ~ .,
              dataset)

#Visualising the Linear Regression result
#install.packages('ggplot2')
library(ggplot2)

ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red')+
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue')+
  ggtitle('Truth or Bluff (Linear Regression)')+
  xlab('Level')+
  ylab('Salary')

#Visualising the Polynomial Regression result
#install.packages('ggplot2')
library(ggplot2)

ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red')+
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue')+
  ggtitle('Truth or Bluff (Polynomial Regression)')+
  xlab('Level')+
  ylab('Salary')

# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
 # library(ggplot2)
 # x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
 # 
 # ggplot() +
 #   geom_point(aes(x = dataset$Level, y = dataset$Salary),
 #              colour = 'red') +
 #   geom_line(aes(x = x_grid, y = predict(poly_reg,
 #                                         newdata = data.frame(Level = x_grid,
 #                                                              Level2 = x_grid^2,
 #                                                              Level3 = x_grid^3,
 #                                                              Level4 = x_grid^4))),
 #             colour = 'blue') +
 #   ggtitle('Truth or Bluff (Polynomial Regression)') +
 #   xlab('Level') +
 #   ylab('Salary')

# Predicting a new result with Linear Regression
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predicting a new result with Polynomial Regression
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                             Level2 = 6.5^2,
                             Level3 = 6.5^3,
                             Level4 = 6.5^4))

File: Polynomial Regression/polynomial_regression_saurabh.py
# Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#Set working directory & import data
dataset = pd.read_csv("Position_Salaries.csv")
X=dataset.iloc[:,1:2].values    #Column index 2 is not included, it is just written to make X as matrix.
Y=dataset.iloc[:,2].values

# Spliting the dataset into Training set & Test set
#We dont have enough data to split it into train and test set
"""from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)"""


#Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,Y)

#Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg =  PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly,Y)

#Visualising the Linear Regression result
plt.scatter(X, Y, color = 'red')
plt.plot(X,lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

#Visualising the Polynomial Regression result
#X_grid = np.arange(min(X), max(X), 0.1)    #To make curve better, smoother, higher resolution
#X_grid = X_grid.reshape(len(X_grid), 1)
#plt.plot(X_grid,lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.scatter(X, Y, color = 'red')
plt.plot(X,lin_reg_2.predict(poly_reg.fit_transform(X)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

#Predicting a new result with Linear Regression
#lin_pred = lin_reg.predict(X)
lin_reg.predict(6.5)

#Predicting a new result with Polynomial Regression
#poly_pred = lin_reg_2.predict(poly_reg.fit_transform(X))
lin_reg_2.predict(poly_reg.fit_transform(6.5))


